design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1900 @@
+- `spec.desiredState`: the desired state, which can be one of `Active` or
+  `Inactive`.
+- `spec.installJobRef`: a reference to the `Job` that was used to unpack the
+  package. The `PackageRevision` must be able to access it to create the objects
+  it installs.
+- `spec.revision`: the number revision that this `PackageRevision` is for its
+  parent `Package`. This is a strictly incrementing value set by the `Package`
+  controller. If an existing `PackageRevision` is transitioned from `Inactive`
+  to `Active`, its revision number will be set to one more than the greatest
+  revision number of all existing `PackageRevision`s for the `Package`. This
+  number is used by the `Package` controller to garbage collect old
+  `PackageRevision`s.
+- `spec.icon`: this replaces the `spec.icons` field and serves as a string
+  representing the SVG icon for a `PackageRevision`. See
+  [#1119](https://github.com/crossplane/crossplane/issues/1119) for more
+  information.
+
+Fields to be modified:
+
+- `spec.dependsOn`: this field is intended to be used for resolving
+  dependencies, but in practice it is only used to make sure that required CRDs
+  are present and that the controller in the package is given RBAC to manage
+  them. It is currently a slice of `PackageInstallSpec` (which would be
+  `PackageSpec` in the new vernacular). In the future, we want the ability to
+  resolve dependencies on either an explicit package (specified by the name of
+  the image), or on CRD(s). However, a package specifying fields like the
+  `imagePullSecrets` for a dependency in its manifests does not make sense
+  because the package author cannot make assumptions about a secret that will be
+  present in the cluster in which it is installed. While out of scope for Phase
+  1, a better model would be to pass any `imagePullSecrets` from the parent
+  `Package` down to any dependencies it requires. The same can be said for
+  `ServiceAccountOptions` and `ImagePullPolicy`. If a user wanted to make it
+  such that dependencies used different configuration than their parent
+  `Package`, they could install the dependency `Package` manually first, then
+  install the parent, which would be have its dependency satisfied by the
+  already extant child. For these reasons, this field will be updated to be of
+  type `[]Dependency` which has mutually exclusive fields `package` and `crd`.
+  The `crd` field should always be a GVK, and the `package` field should be an
+  image reference, with a semver range.
+
+*Note: `spec.dependsOn` will also no longer be used to drive RBAC generation.
+Declaring a dependency means that the package requires that the dependency
+(whether another package or a CRD) exists in the cluster. It does not indicate
+that the package wants to be able to reconcile the type. Packaged controllers
+may only reconcile types that they install.*
+
+Fields to be removed:
+
+- `spec.packageType`: currently is one of `Provider`, `Stack`, `Application`, or
```

Comment:
IIRC there are still a use cases to distinguish providers from stacks.

Reply:
hmm I am not familiar with this use case. Can you provide more information?

Reply:
@hasheddan When pushing the package to a registry we'd like to categorize the package type as a provider or a composition (for filtering, searching etc.). With the new format can that be determined in a different way?

Reply:
@prasek gotcha! Looked into this a little more, could you confirm my conclusions:
- A package of type `Provider` can bring a controller and CRDs and nothing else
- A package of type `Stack` (or did we want to call that a `Composition`?) can bring InfraDef / InfraPub / Compositions and nothing else

Reply:
@thephred @rathpc - can one of you confirm what requirements we have here for this type?

Reply:
@lukeweber this would not be a new type, just a `PackageRevision` where the `packageType: Composition`